### ISSUE TITLE
ENH: Allow `sphx-glr-multi-img` images to span multiple lines

### ISSUE THREAD

**Issue by tsbinns** [sphinx-gallery/sphinx-gallery#1371]
Discussed with @larsoner.

When plotting multiple figures from the same cell, the figures are placed in the html as `sphx-glr-multi-img` images in a `sphx-glr-horizontal` list. Although each figure is stored in a new list element, if the figures are wide, they are shrunk down regardless of the specified figure sizes and placed on (what is essentially) the same line, which can make them illegible, e.g.: https://mne.tools/mne-connectivity/dev/auto_examples/decoding/cohy_decomposition_plotting.html#visualising-the-filters

It would be nice to have some sort of option that if multiple figures are shown from the same cell, the list elements containing each of the figures would be placed beneath one another. E.g.,

Before:
![image](https://github.com/user-attachments/assets/10b8eab1-ac66-434e-9d96-c2368f8bb92f)

After:
![image](https://github.com/user-attachments/assets/34371b4a-38e0-43c6-b807-8d553a3c2814)

Other ideas are also welcome. If people think this is a sensible feature I am happy to write a PR (though I wouldn't have the time until later in the year).

**Comment by larsoner** [comment]
Agreed this would be nice. I know I've wanted something like this before and presumably @drammock has, too, when working on tutorials.

**Comment by drammock** [comment]
> Agreed this would be nice. I know I've wanted something like this before and presumably @drammock has, too, when working on tutorials.

yep, just been too busy to request/implement it...

**Comment by lucyleeow** [comment]
Naive question, is there a clever CSS/javascript way to detect image size/ratio and automatically decide on best layout?

**Comment by tsbinns** [comment]
I have some time to look into this now; wanted to get some thoughts on the best way to proceed.

So currently, the max image width in `sg_gallery.css` for the `multi-img` class is 42%, which gives the 2 columns of images (compared to 100% max width for `single-img` class).

> Naive question, is there a clever CSS/javascript way to detect image size/ratio and automatically decide on best layout?

I'm also not sure about this. A simple approach would just be to allow `multi-img` max width to also be 100%, which would prevent things being shrunk.

However, I'm guessing it would be preferred if the default behaviour of 42% width remains and a new option for allowing wider `multi-img` images is added? In this case, do you think it's better if the option toggles between 42% and 100% max width, or should the user be allowed to set any max width for more flexibility? Or, does someone else have another suggestion they think might work better?

From a quick look, I'm not certain where the max width CSS parameter could be intercepted and modified on a per-example basis, but I wanted to get your thoughts on whether this idea is sensible before digging in too much.

**Comment by larsoner** [comment]
> From a quick look, I'm not certain where the max width CSS parameter could be intercepted and modified on a per-example basis

On a *per-block* basis would be tough I think? But on a *per-example* basis though I think it's not so bad because you can add a per-page CSS.

But to me rather than this CSS stuff I think it's easiest to add a config var that makes SG change the RST to contain multiple `single-img` rather than one `multi-img`. If users want:

1. Global no-width-limit: change the CSS `max-width` for `multi-img`
2. A single example no-width-limit: change set the `# sphinx_gallery_multi_image = single` *somewhere in the file*
3. A single block with an example no-width-limit: set the `# sphinx_gallery_multi_image_block = single` *within that block*

I think something like this would work?

**Comment by tsbinns** [comment]
Okay, so if I understand correctly, you'd recommend an option to switch from one `multi-img` to multiple `single-img`. So I should focus on something like point 2 and leave 1 & 3 out, or include all to give users more choice?

**Comment by larsoner** [comment]
(1) is really just a doc change hopefully. Implementing either (2) or (3) would make the other easy enough to implement. So I'd say start with (2) or (3), see if it's trivial to then add (3) or (2), respectively, and if you're motivated to try and document (1) feel free :slightly_smiling_face:

**Comment by tsbinns** [comment]
Had some time to look into this and was hoping to get some advice:

From what I see, `multi-img`s get created in `matplotlib_scraper()` (L212-215):
https://github.com/sphinx-gallery/sphinx-gallery/blob/a3a8de1034839a164d8da0eb057e4a9612ec3ab2/sphinx_gallery/scrapers.py#L209-L219

My first thought was to pass something like `file_conf` to the scraper, but that gives the following error: `TypeError: MatplotlibFormatScraper.__call__() takes 4 positional arguments but 5 were given` and it looks like `kwargs` are reserved for `savefig()` params.

I suppose if something like `# sphinx_gallery_multi_image_block = "single"` was present for a block, then in the `execute_code_block()` function this could be inserted into the `block_vars` param which is passed to `save_figures()`, which in turn is passed to the scraper.

Within the scaper, you could substitute  `single-img` for `multi-img` only if `block_vars["multi_image_block"] != "single"`, e.g.:
```
...
elif len(image_rsts) > 1: 
    if block_vars["multi_image_block"] != "single":
        image_rsts = [ 
            re.sub(r":class: sphx-glr-single-img", ":class: sphx-glr-multi-img", image) 
            for image in image_rsts
        ]
...
```

If there were no block-specific settings but rather a file-wide `# sphinx_gallery_multi_image = "single"`, then `# sphinx_gallery_multi_image_block = "single"` could get added for each code block in the `execute_code_block()` function.

Does this seem like a reasonable implementation?

One thing I imagine is if the user defines a custom scraper, it would need to include logic to handle the `sphinx_gallery_multi_image_block` option, otherwise the feature would not work. Is this considered acceptable behaviour, or do all config options need to work regardless of the scraper being used?

**Comment by lucyleeow** [comment]
> My first thought was to pass something like file_conf to the scraper, but that gives the following error: TypeError: MatplotlibFormatScraper.__call__() takes 4 positional arguments but 5 were given and it looks like kwargs are reserved for savefig() params.

Can you expand on how you got that error?

I tried adding a `file_conf` parameter to the matplotlib scraper and it was fine?

**Comment by tsbinns** [comment]
Ah sorry, I just checked again and there was something I had overlooked.
If people are happy for `file_conf` to be passed through to the scaper, I could open a PR handling a file-wide option.
Could then also look into the block-wise option case there.

**Comment by lucyleeow** [comment]
@tsbinns is this fixed by #1384 ?

**Comment by tsbinns** [comment]
Yes, sorry, thought I had closed this already!
